presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,39 +1594,197 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
+            return true;
+        }
+        return false;
+    }
 
-        if (end - start < 3) {
-            // A path with less than 3 distinct points is not valid for calculating an area
-            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, "Polygon is not valid: a loop contains less then 3 vertices.");
+    private static boolean edgeCrossesLongitude(double edgeStart, double edgeEnd, double longitude)
+    {
+        if (Double.compare(edgeStart, longitude) == 0) {
+            return true;
+        }
+        if (Double.compare(edgeEnd, longitude) == 0) {
+            return true;
         }
 
-        Point point = new Point();
+        double start = min(edgeStart, edgeEnd);
+        double end = max(edgeStart, edgeEnd);
 
-        // Initialize the calculator with the last point
-        polygon.getPoint(end - 1, point);
-        SphericalExcessCalculator calculator = new SphericalExcessCalculator(point);
+        if (haveSameLongitude(edgeEnd, longitude)) {
+            return false;
+        }
+        if (haveSameLongitude(edgeStart, longitude)) {
+            return true;
+        }
 
-        for (int i = start; i < end; i++) {
-            polygon.getPoint(i, point);
-            calculator.add(point);
+        double distanceVia0 = end - start;
+        double distanceVia180 = start - MIN_LONGITUDE + MAX_LONGITUDE - end;
+        if (distanceVia0 < distanceVia180) {
+            return liesInRange(longitude, start, end);
         }
+        return liesInRange(longitude, MIN_LONGITUDE, start) || liesInRange(longitude, end, MAX_LONGITUDE);
+    }
 
-        return calculator.computeSphericalExcess();
+    private static boolean isVerticalEdge(Point edgeStart, Point edgeEnd)
+    {
+        return haveSameLongitude(edgeStart.getX(), edgeEnd.getX());
     }
 
-    private static class SphericalExcessCalculator
+    private static boolean edgeIntersectsTestEdge(Point edgeStart, Point edgeEnd, Point testEdgeStart, Point testEdgeEnd)
+    {
+        // if this edge never crosses the longitude of the test edge, they can't possibly intersect
+        if (!edgeCrossesLongitude(edgeStart.getX(), edgeEnd.getX(), testEdgeStart.getX())) {
+            return false;
+        }
+
+        // otherwise, calculate the edge's latitude at the point's longitude
+        double edgeLatitude = SphericalCalculator.computeExtrapolatedLatitude(edgeStart, edgeEnd, testEdgeStart.getX());
+        boolean inYRange = liesInRange(edgeLatitude, testEdgeStart.getY(), testEdgeEnd.getY()) && testEdgeStart.getY() < edgeLatitude;
+        boolean inXRangeExclusive = (Double.compare(edgeStart.getX(), testEdgeStart.getX()) != 0); //&& Double.compare(edgeEnd.getX(), testEdgeStart.getX()) >= 0);
+        return inXRangeExclusive && inYRange;
+    }
+
+    private static boolean longitudeGrazesVertex(double longitude, Point pointPreVertex, Point vertex, Point pointPostVertex)
```

Comment:
I think you can avoid this function entirely by using the edge-crossing rules (http://geomalgorithms.com/a03-_inclusion.html#Edge-Crossing-Rules).  In the planar case, where we use a ray from the point to `x -> +infinity`:
1. Horizontal edges are ignored,
2. Upward edges include starting point and exclude ending point,
3. Downward edges exclude starting point and include ending point,
4. Edges that the point is on are ignored.

Since you are checking a ray to the north pole, I think we have to rotate this 90-degrees counter-clockwise.  I think this means it becomes (please check!):
1. Vertical edges are ignored,
2. Westward edges include starting point and exclude ending point,
3. Eastward edges exclude starting point and include ending point,
4. Edges that the point is on are ignored.

This should simplify the logic for the algorithm considerably, primarily leaving the logic dealing with the antimeridian.